engine/src/main/scala/cromwell/services/KeyValueServiceActor.scala

```diff
@@ -0,0 +1,69 @@
+package cromwell.services
+
+import akka.actor.{Actor, ActorRef, Props}
+import com.typesafe.config.Config
+import cromwell.backend.BackendJobDescriptor
+import cromwell.engine.db.DataAccess
+import cromwell.services.KeyValueServiceActor._
+
+import scala.concurrent.Future
+import scala.util.{Failure, Success}
+
+object KeyValueServiceActor {
+  trait KvMessage
+  trait KvAction extends KvMessage
+  trait KvResponse extends KvMessage
+
+  case class KvPut(jobDescriptor: BackendJobDescriptor, key: String, value: String) extends KvAction
+  case class KvGet(jobDescriptor: BackendJobDescriptor, key: String) extends KvAction
+
+  case class KvPair(jobDescriptor: BackendJobDescriptor, key: String, value: Option[String]) extends KvResponse
+  case class KvFailure(action: KvAction, failure: Throwable) extends KvResponse
+  case class KvKeyLookupFailed(action: KvGet) extends KvResponse
+  case class KvPutSuccess(action: KvPut) extends KvResponse
+
+  def props(dataAccess: DataAccess, serviceConfig: Config, config: Config) = {
+    Props(KeyValueServiceActor(dataAccess, serviceConfig, config))
+  }
+}
+
+case class KeyValueServiceActor(dataAccess: DataAccess,
+                                serviceConfig: Config,
+                                entireConfig: Config) extends Actor {
+  private implicit val ec = context.dispatcher
+
+  def receive = {
+    case action: KvGet => respond(sender(), action, doGet(action))
+    case action: KvPut => respond(sender(), action, doPut(action))
+  }
+
+  private def respond(replyTo: ActorRef, action: KvAction, future: Future[KvResponse]): Unit = {
+    future.onComplete {
+      case Success(x) => replyTo ! x
+      case Failure(ex) => replyTo ! KvFailure(action, ex)
+    }
+  }
+
+  private def doPut(put: KvPut): Future[KvResponse] = {
+    dataAccess.upsertExecutionInfo(
+      put.jobDescriptor.descriptor.id,
+      put.jobDescriptor.key,
+      Map(put.key -> Option(put.value))
+    ).map(_ => KvPutSuccess(put))
+  }
+
+  private def doGet(get: KvGet): Future[KvResponse] = {
+    val executionInfos = dataAccess.getExecutionInfos(
+      get.jobDescriptor.descriptor.id,
+      get.jobDescriptor.key.call,
+      get.jobDescriptor.key.attempt
+    )
+
+    executionInfos map {
+      case infos => infos.find(_.key == get.key) match {
```

Comment:
Yeah I guess so... let me rebase on top of the `deadlocks_and_upserts` first so I don't create even more merge conflicts